### PR TITLE
refactor(code) Major refactor of Account class

### PR DESCRIPTION
**Pull request by shitwolfymakes**
## Summary
This PR is major refactor of the `Account` class. It preserves as much of the original interface as possible, and what breaking changes were been made have been fixed.

### API Changes
- Added function `SetCredits()`, with 100% test coverage
- Added function `AddHistory()`, with 100% test coverage
- Renamed function `CrewSalariesOwed()` to `OverdueCrewSalaries()` to increase clarity, added 100% test coverage
- Added function `SetOverdueCrewSalaries()`, with 100% test coverage
- Renamed function `PaySalaries()` to `PayOverdueCrewSalaries()` to increase clarity, added 100% test coverage
- Renamed function `MaintenanceDue()` to `OverdueMaintenance()` to increase clarity, added 100% test coverage
- Added function `SetOverdueMaintenance()`, with 100% test coverage
- Renamed function `PayMaintenance()` to `PayOverdueMaintenance()` to increase clarity, added 100% test coverage

### Implementation Changes
- Heavily refactored `Step()` to increase clarity, added 100% test coverage
  - Added `Receipt` struct to handle return data in Step utility functions (Could easily replaced with `std::tuple`, but that would increase output size and compile time much more than a custom struct definition)
  - Added `PayCrewSalaries()` utility function
  - Added `PayMaintenance()` utility function
  - Added `PayMortgages()` utility function
  - Added `PayFines()` utility function
  - Added `GenerateMissedPaymentLogs()` utility function
  - Added `UpdateMortgages()` utility function
  - Added `CalculateNetWorth()` utility function
  - Added `UpdateCreditScore()` utility function
  - Added `AnyPaymentsMade()` utility function
  - Added `GeneratePaymentLogs()` utility function
  - Added `GetTypesPaid()` utility function
- Renamed member `crewSalariesOwed` to `overdueCrewSalaries` to increase clarity
- Renamed member `maintenanceDue` to `overdueMaintenance` to increase clarity

### Miscellaneous Changes
- In the header file, function comments have been added for all members, and describe *how they should be used* by clients of the interface.
- In the source file, function comments of the type mentioned above have been removed to avoid duplication issues. Some function comments have been added, describing *how the function should be implemented*. This keeps unnecessary implementation details out of the interface.

### Automated Tests Added
See above. Additionally, 100% test coverage has been added for all public functions except `Save()`, `Load()`, and `Prequalify()`.

### CODE REVIEW

**Comment by quyykk** [comment]
I'm not sure splitting Step just for unit testing is a good idea. Can't you test the effects of Step, i.e. whether it correctly deducts costs and/or adds your salary? Because the code calls Step and not the intermediate functions, so either you Step should be tested if that makes sense. Thoughts?

**Comment by shitwolfymakes** [comment]
> I'm not sure splitting Step just for unit testing is a good idea. Can't you test the effects of Step, i.e. whether it correctly deducts costs and/or adds your salary? Because the code calls Step and not the intermediate functions, so either way Step should be tested if that makes sense. Thoughts?

At this point this is no longer a refactor of Step alone, but rather a refactor of the whole Account class. Test coverage will include everything I touch

**Comment by quyykk** [comment]
I think it's not a good idea to be adding such a large amount of public functions. The code is too abstracted away and at least for me, it's harder to follow along because I need to go through multiple functions mentally.

IMO you shouldn't be adding public functions to the interface if they're not needed, and just test the public interface.

**Comment by shitwolfymakes** [comment]
What public functions are you referring to?

**Comment by shitwolfymakes** [comment]
@quyykk bump

**Comment by quyykk** [comment]
> What public functions are you referring to?

@shitwolfymakes all the ones that are added. At least to me, it makes reading and understanding the code much more complex.

**Comment by shitwolfymakes** [comment]
Walk me through your thought process? The only new public functions are required to be able to test the interface.

If you mean that the previous implementation of `Step()` was less complex, that is not the case. The old implementation was 142 lines long, with a cognitive complexity of 57. This refactor brings it down to 20 lines, with a cognitive complexity of >= 5.

If you're referring to the new Bill struct, I agree it adds complexity. However, using that avoids having to link the tuple headers which saves on compile time and size. I'd be happy to scrap it and just use tuples if you'd prefer.

File: source/Account.cpp [hunk 1]
```diff
@@ -115,29 +113,87 @@ void Account::Save(DataWriter &out) const
...
+
+
+int64_t Account::NetWorth() const
+{
+	return history.empty() ? 0 : history.back();
+}
+
+
+
+void Account::AddHistory(int64_t amount) {
```
Comment (RisingLeaf) [inline source/Account.cpp:158]:
```suggestion
void Account::AddHistory(int64_t amount)
{
```

File: source/Account.cpp [hunk 2]
```diff
@@ -115,29 +113,87 @@ void Account::Save(DataWriter &out) const
...
+	history.push_back(amount);
+	if(history.size() > HISTORY)
```
Comment (RisingLeaf) [inline source/Account.cpp:160]: should be named MAX_HISTORY_SIZE or somth like that

File: source/Account.cpp [hunk 3]
```diff
@@ -115,29 +113,87 @@ void Account::Save(DataWriter &out) const
...
+
+
 void Account::PayExtra(int mortgage, int64_t amount)
 {
-	if(static_cast<unsigned>(mortgage) >= mortgages.size() || amount > credits
-			|| amount > mortgages[mortgage].Principal())
+	if(static_cast<unsigned>(mortgage) >= mortgages.size()
+		|| amount > credits
+		|| amount > mortgages[mortgage].Principal())
+	{
```
Comment (RisingLeaf) [inline source/Account.cpp:194]:
```suggestion
```

File: source/Account.cpp [hunk 4]
```diff
@@ -115,29 +113,87 @@ void Account::Save(DataWriter &out) const
...
 		return;
+	}
```
Comment (RisingLeaf) [inline source/Account.cpp:196]:
```suggestion
```

File: source/Account.cpp [hunk 5]
```diff
@@ -150,149 +206,85 @@ void Account::PayExtra(int mortgage, int64_t amount)
...
-	}
-	return out.str();
+int64_t Account::OverdueCrewSalaries() const
+{
+	return overdueCrewSalaries;
+}
+
+
+
+void Account::SetOverdueCrewSalaries(int64_t value) {
```
Comment (RisingLeaf) [inline source/Account.cpp:249]:
```suggestion
void Account::SetOverdueCrewSalaries(int64_t value)
{
```

File: source/Account.cpp [hunk 6]
```diff
@@ -150,149 +206,85 @@ void Account::PayExtra(int mortgage, int64_t amount)
...
+
+
+int64_t Account::OverdueMaintenance() const
+{
+	return overdueMaintenance;
+}
+
+
+
+void Account::SetOverdueMaintenance(int64_t value) {
```
Comment (RisingLeaf) [inline source/Account.cpp:274]:
```suggestion
void Account::SetOverdueMaintenance(int64_t value)
{
```

File: source/Account.cpp [hunk 7]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	out << GeneratePaymentLogs(&receipts);
+	return out.str();
 }
 
 
 
-int64_t Account::MaintenanceDue() const
-{
-	return maintenanceDue;
+std::vector<Receipt> Account::PayBills(int64_t salaries, int64_t maintenance) {
```
Comment (RisingLeaf) [inline source/Account.cpp:348]:
```suggestion
std::vector<Receipt> Account::PayBills(int64_t salaries, int64_t maintenance)
{
```

File: source/Account.cpp [hunk 8]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
 
 
 
-// Add a new mortgage for the given amount, with an interest rate determined by
-// your credit score.
-void Account::AddMortgage(int64_t principal)
-{
-	mortgages.emplace_back(principal, creditScore);
-	credits += principal;
+Receipt Account::PayMortgages() {
```
Comment (RisingLeaf) [inline source/Account.cpp:427]:
```suggestion
Receipt Account::PayMortgages()
{
```

File: source/Account.cpp [hunk 9]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	// This function needs to preserve:
+	//    1. The number of credits paid towards ALL mortgages
+	//    2. Whether ANY mortgage was not paid in full
+	Receipt receipt;
+	if(mortgages.empty()) return receipt;
+
+	for(Mortgage &mortgage : mortgages)
+	{
+		if(!(mortgage.Type() == "Mortgage")) continue;
```
Comment (RisingLeaf) [inline source/Account.cpp:436]:
```suggestion
		if(!(mortgage.Type() == "Mortgage"))
			continue;
```

File: source/Account.cpp [hunk 10]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	return receipt;
 }
 
 
 
-// Add a "fine" with a high, fixed interest rate and a short term.
-void Account::AddFine(int64_t amount)
-{
-	mortgages.emplace_back(amount, 0, 60);
+Receipt Account::PayFines() {
```
Comment (RisingLeaf) [inline source/Account.cpp:456]:
```suggestion
Receipt Account::PayFines()
{
```

File: source/Account.cpp [hunk 11]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
 
 
 
-// Check how big a mortgage the player can afford to pay at their current income.
-int64_t Account::Prequalify() const
-{
-	double payments = 0.;
-	int64_t liabilities = 0;
-	for(const Mortgage &mortgage : mortgages)
+const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const {
```
Comment (RisingLeaf) [inline source/Account.cpp:485]:
```suggestion
const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const
{
```

File: source/Account.cpp [hunk 12]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
-	// Put a limit on new debt that the player can take out, as a fraction of
-	// their net worth, to avoid absurd mortgages being offered when the player
-	// has just captured some very lucrative ships.
-	return max<int64_t>(0, min(
-		NetWorth() / 3 + 500000 - liabilities,
-		Mortgage::Maximum(YearlyRevenue(), creditScore, payments)));
+	if(!receipts->at(1).paidInFull)
+	{
+		log << "You could not pay all your maintenance costs. ";
+	}
```
Comment (RisingLeaf) [inline source/Account.cpp:496]:
```suggestion
		log << "You could not pay all your maintenance costs. ";
```

File: source/Account.cpp [hunk 13]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+
+	if(Mortgages().size() > 0)
+	{
+		if(!receipts->at(2).paidInFull || !receipts->at(3).paidInFull)
+		{
+			log << "You missed a mortgage payment. ";
+		}
+	}
```
Comment (RisingLeaf) [inline source/Account.cpp:504]:
```suggestion
		if(!receipts->at(2).paidInFull || !receipts->at(3).paidInFull)
			log << "You missed a mortgage payment. ";
```

File: source/Account.cpp [hunk 14]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
 
 
-// Get the total amount owed for "Mortgage", "Fine", or both.
-int64_t Account::TotalDebt(const string &type) const
-{
-	int64_t total = 0;
-	for(const Mortgage &mortgage : mortgages)
-		if(type.empty() || mortgage.Type() == type)
-			total += mortgage.Principal();
+void Account::UpdateCreditScore(std::vector<Receipt> *receipts) {
```
Comment (RisingLeaf) [inline source/Account.cpp:531]:
```suggestion
void Account::UpdateCreditScore(std::vector<Receipt> *receipts)
{
```

File: source/Account.cpp [hunk 15]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	// makes you at least as credit-worthy as someone who pays debts on time.)
+	bool missedPayment = false;
+	for(Receipt receipt : *receipts)
+	{
+		if(!receipt.paidInFull)
+		{
+			missedPayment = true;
+			break;
+		}
+	}
```
Comment (RisingLeaf) [inline source/Account.cpp:543]:
```suggestion
		if(!receipt.paidInFull)
		{
			missedPayment = true;
			break;
		}
```

File: source/Account.cpp [hunk 16]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	creditScore = max(200, min(800, creditScore + (missedPayment ? -5 : 1)));
+}
 
-	return total;
+
+
+bool Account::AnyPaymentsMade(std::vector<Receipt> *receipts) {
```
Comment (RisingLeaf) [inline source/Account.cpp:549]:
```suggestion
bool Account::AnyPaymentsMade(std::vector<Receipt> *receipts)
{
```

File: source/Account.cpp [hunk 17]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	for(Receipt receipt : *receipts)
+	{
+		if(receipt.creditsPaid > 0)
+			return true;
+	}
```
Comment (RisingLeaf) [inline source/Account.cpp:554]:
```suggestion
		if(receipt.creditsPaid > 0)
			return true;
```

File: source/Account.cpp [hunk 18]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+	return false;
+}
+
+
+
+string Account::GeneratePaymentLogs(std::vector<Receipt> *receipts) {
```
Comment (RisingLeaf) [inline source/Account.cpp:560]:
```suggestion
string Account::GeneratePaymentLogs(std::vector<Receipt> *receipts)
{
```

File: source/Account.cpp [hunk 19]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
-int64_t Account::Prequalify() const
-{
-	double payments = 0.;
-	int64_t liabilities = 0;
-	for(const Mortgage &mortgage : mortgages)
+const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const {
+	ostringstream log;
+
+	if(!receipts->at(0).paidInFull)
 	{
```
Comment (RisingLeaf) [inline source/Account.cpp:489]:
```suggestion
```

File: source/Account.cpp [hunk 20]
```diff
@@ -329,116 +321,298 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
-		payments += mortgage.PrecisePayment();
-		liabilities += mortgage.Principal();
+		log << "You could not pay all your crew salaries. ";
 	}
```
Comment (RisingLeaf) [inline source/Account.cpp:491]:
```suggestion
```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
style checkup

File: source/Account.cpp (hunk 3 above)
Comment (shitwolfymakes) [reply source/Account.cpp:195]: @quyykk I disagree with removing the curly braces here. Thoughts?

Comment (RisingLeaf) [reply source/Account.cpp:195]: b- but less line

Comment (shitwolfymakes) [reply source/Account.cpp:195]: Lmao, but it look more gross your way

File: source/Account.cpp (hunk 17 above)
Comment (shitwolfymakes) [reply source/Account.cpp:557]: For loops should always be enclosed with curly braces

Comment (RisingLeaf) [reply source/Account.cpp:557]: style guide disagrees

Comment (shitwolfymakes) [reply source/Account.cpp:551]: source?

Comment (RisingLeaf) [reply source/Account.cpp:551]: you can look at the style guide in the wiki, I am going to sleep now

Comment (shitwolfymakes) [reply source/Account.cpp:551]: Says you *can*, not that you must. And for a for loop, even if it's legal C++ I'm gonna keep it encased in curly braces

**Comment by shitwolfymakes** [comment]
Updating from master broke this somehow, looks like CI errors and not any of my recent commits. Can a rerun be initiated?

**Comment by shitwolfymakes** [comment]
Okay, looks like the failing tests were caused by committing @RisingLeaf's suggestions. Will double check everything

**Comment by shitwolfymakes** [comment]
Formatting issues fixed and all tests are passing again. Not sure what broke but it's fixed now. Ready for review

**Comment by quyykk** [comment]
> If you mean that the previous implementation of Step() was less complex, that is not the case. The old implementation was 142 lines long, with a cognitive complexity of 57. This refactor brings it down to 20 lines, with a cognitive complexity of >= 5.

I disagree. The functions might be smaller but the code is now spread out into multiple small-ish functions and it makes it hard to follow the code (due to needing to jump around so much). Abstractions are usually a good idea but if you just have a couple of lines per function? Idk, at least to me reading the new code is harder because it seems harder to understand the whole class.

> If you're referring to the new Bill struct, I agree it adds complexity. However, using that avoids having to link the tuple headers which saves on compile time and size. I'd be happy to scrap it and just use tuples if you'd prefer.

I don't mind the struct, and it's definitely better than a tuple imo.

**Comment by shitwolfymakes** [comment]
Glad we're on the same page about the Bill struct. I'll see where I can make some changes on the other stuff.

**Comment by shitwolfymakes** [comment]
So I've looked over the code and clearly I'm not seeing what you're seeing. What functions are you having a hard time following, and how are they harder to follow than the original, ~200 line function?

**Comment by quyykk** [comment]
Take Step for example. Sure it's a long function, but that's because it's doing a lot of stuff. But I can go through it and it's fairly understandable.

With your changes though, the function is really short but it calls a lot of functions who in turn call more functions. This is hard to keep track of when reading the code because I need to jump back and forth to various functions to understand what Step is doing. At least IMO

**Comment by shitwolfymakes** [comment]
I disagree. There is an upfront cost in familiarizing yourself with the new code, but every function called is well named, and encapsulates each discrete step(heh) that the Step function walks through. Step doesn't need to know how the bills are paid, just that it has receipts to hand off to the log writer. SRP.

At this point, I see no technical reason why this can't be merged. The code is different now, but it's objectively easier to maintain going forward and the tests are included to ensure each part is working properly when changes are made in the future.

**Comment by Lorantine** [comment]
Regarding the last two comments between quyykk and wolfy, I'd agree with the latter, given it's done properly.

A method with hundreds of lines of code that does a lot becomes a big reading list where you can get lost in, while splitting parts into logical smaller ones makes it easier to see. Sure you need to jump to them a bit to see what they do, but when you basically "reviewed all those small ones and know they work", you're left with reading the short(er) main method and seeing mostly just a few method calls and small bits of code you can read through in a matter of seconds. "Oh, this part gets this, then sets this, and does this" (3 methods were used, each being ~50 lines but you reviewed these before and know they work well).

guyykk's one is more of a preference, though I wouldn't recommend it for code maintainability. (At work I have some classes with 6-10k+ lines and some giant methods of a few thousand lines, and there you'd wish it was split more... It just becomes too much.)

**Comment by Lorantine** [comment]
Anyway, with all that said, I took a look:
I just want some more comments in 'test_account.cpp' for more transparency what exactly happens and why.
When that's done, I, in my finite C++ wisdom (none, that is, but 4 years of C# wisdom), would give this my "ok" for merging.
Though an ok from a proper code reviewer is still warranted then.

**Comment by shitwolfymakes** [comment]
@Lorantine Lemme know if the new comments helps

File: source/Account.h [hunk 21]
```diff
@@ -21,75 +21,181 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
 public:
-	// Load or save account data.
+	// Load account information from a data file (saved game or starting conditions).
 	void Load(const DataNode &node, bool clearFirst);
+
+	// Write account information to a saved game file.
 	void Save(DataWriter &out) const;
 
-	// Get or change the player's credits.
+	// Get much the player currently has in the bank.
```
Comment (petervdmeer) [inline source/Account.h:48]:
```suggestion
	// Get how much credits the player currently has in the bank.
```

File: source/Account.h [hunk 22]
```diff
@@ -21,75 +21,181 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include <cstdint>
 #include <map>
 #include <string>
+#include <tuple>
 #include <vector>
 
 class DataNode;
 class DataWriter;
 
+// A utility stucture to return multpile values needed during processing
```
Comment (petervdmeer) [inline source/Account.h:30]:
```suggestion
// A utility structure to return multiple values needed during processing
```

File: source/Account.cpp [hunk 23]
```diff
@@ -329,116 +324,300 @@ void Account::SetSalaryIncome(const string &name, int64_t amount)
...
+		return out.str();
+
+	out << GeneratePaymentLogs(&receipts);
+	return out.str();
 }
 
 
 
-void Account::PaySalaries(int64_t amount)
+std::vector<Receipt> Account::PayBills(int64_t salaries, int64_t maintenance)
```
Comment (petervdmeer) [inline source/Account.cpp:351]: This function name seems to be a little too generic, and a little bit off (at least for my native country/language):
- Salary payments are accompanied by a payslip. Only contractors typically send a bill to get their salary. Others just get a payslip from their employer without sending any bill by themselves.
- Fines are just sent as fine (we don't call them bill, we just call them fine)
- Mortgage payments are typically automatically deducted from bank-accounts. Banks don't send a bill, but they do send receipts (and when needed a reminder).

I would recommend to move the content from this function into Step(...) It is only 4 functions and all those four functions have very clear, specific, well-defined names.

(I believe that moving the content of this function into Step() and then dropping this function also addresses some of the scrolling concerns which quyykk has.)

**Review by petervdmeer — APPROVED** [review]
> Take Step for example. Sure it's a long function, but that's because it's doing a lot of stuff. But I can go through it and it's fairly understandable.
> 
> With your changes though, the function is really short but it calls a lot of functions who in turn call more functions. This is hard to keep track of when reading the code because I need to jump back and forth to various functions to understand what Step is doing. At least IMO

I feel that the new code is easier to read and it seems to be better testable, and better tested. I do also understand some of quyykks concern on scrolling around and I feel that PayBills() is an example of a function where it is not immediately clear what it does and what it doesn't do (thus needing to scroll to the function and reading its contents instead of just looking at the name). I did make a recommendation on PayBills() on how I feel this can be improved, but the code as-is does already seem to be an improvement over what we had, so I would already be okay with merging this PR as-is. (After someone reviews all the code.)

**Comment by Lorantine** [comment]
For my part (given peter's typo feedback is followed up on), I'm happy with it. The comments do help a decent bit.

File: source/Account.h (hunk 21 above)
Comment (TheGiraffe3) [reply source/Account.h:48]:
```suggestion
	// Get how many credits the player currently has in the bank.
```
Did you mean like this?

**Comment by petervdmeer** [comment]
"Waiting on Op": merge conflicts and some last minor open items.

**Review by petervdmeer — CHANGES_REQUESTED** [review]
Waiting on OP, and merge-conflicts.

File: source/Account.cpp (hunk 23 above)
Comment (shitwolfymakes) [reply source/Account.cpp]: Resolved with 4ce7945

**Review by shitwolfymakes — COMMENTED** [review]
Changes implemented

File: source/Account.cpp (hunk 2 above)
Comment (petervdmeer) [reply source/Account.cpp:161]: Something like MAX_HISTORY_ENTRIES indeed seems more clear than just HISTORY.

Comment (RisingLeaf) [reply source/Account.cpp:161]: also history could be a queue, would fit the nature of adding at the back and removing the front

File: source/Account.cpp [hunk 24]
```diff
@@ -150,322 +219,448 @@ void Account::PayExtra(int mortgage, int64_t amount)
...
-int64_t Account::MaintenanceDue() const
-{
-	return maintenanceDue;
+	return log.str();
 }
 
 
 
-void Account::PayMaintenance(int64_t amount)
+void Account::UpdateMortgages()
```
Comment (petervdmeer) [inline source/Account.cpp:545]: Could we make this part of `PayMortgages()`. It feels very appropriate to just remove it from the list directly after the last full payment is made than in a separate function.

File: source/Account.cpp (hunk 17 above)
Comment (petervdmeer) [reply source/Account.cpp:590]: The text in the "Single-line loop bodies" entry indeed didn't sound very directive, but I'm very sure that the text `"do not need to"` really should be read as `"should not"`. Styleguide entries are not written with the intention of being optional.
I created https://github.com/endless-sky/endless-sky.github.io/pull/48 to make this entry in the styleguide more clear.

File: source/Account.cpp [hunk 25]
```diff
@@ -150,322 +219,448 @@ void Account::PayExtra(int mortgage, int64_t amount)
...
 
 
-int64_t Account::CrewSalariesOwed() const
+const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const
 {
-	return crewSalariesOwed;
-}
+	ostringstream log;
 
+	if(!receipts->at(0).paidInFull)
```
Comment (petervdmeer) [inline source/Account.cpp:528]: Using the index in the vector as a magic number to determine which item we are looking at is not very elegant and maintainable.

Maybe consider putting in the receipt itself what the receipt is for? (So the receipt at position 0 would have a string in it that says "crew salaries", and the receipt at position 1 would have a string that says "maintenance costs".)

File: source/Account.cpp (hunk 24 above)
Comment (shitwolfymakes) [reply source/Account.cpp:545]: Leaving this as a generic function is preferable, as there are at least 3 types of mortgages (Mortgages, Fines, and Debt) stored together by the game engine now that #9286 has been merged.

File: source/Account.cpp (hunk 25 above)
Comment (petervdmeer) [reply source/Account.cpp:528]: No, documenting the magic numbers would not be sufficient. And I'm not looking for complex solutions with maps or tuple headers.
It's just that we have information hidden within the positions of the vector, and that is makes the code less maintainable.

File: source/Account.cpp (hunk 24 above)
Comment (petervdmeer) [reply source/Account.cpp:545]: Removing the Mortgage after it has been fully paid should be about 2 lines of code, within the function where it is relevant, without anyone new to the code needing to do "mental switching" (as quyykk described it) to understand how they relate to each-other.

If we need future leverage with common functions for mortgages, then we can introduce them once we need them.

At this moment I feel that this 4 lines-function is just an example of the earlier remark that the code is split in functions that are too small.

Comment (shitwolfymakes) [reply source/Account.cpp:545]: Can we split the difference, and I remove the call as you're requesting, but leave the utility function in so it's ready to go later down the line?

File: source/Account.cpp (hunk 25 above)
Comment (shitwolfymakes) [reply source/Account.cpp:528]: I don't disagree, magic numbers are something to be avoided most of the time. There's a couple of ways to do what you'd like as I see it:
1. Create a map of Receipt and type (salary, maintenance, mortgages), which requires linking the map headers and defeats the original purpose of using the custom struct, or
2. Add a type enum to the struct, then loop through the receipt vector and check for the type to operate on, which adds an unnecessary loop and is very amateurish, despite the performance impact from a loop that small and run once per time step is functionally zero.

I'm happy to go with option 1, but I do think this is one of the rare cases where magic numbers are inoffensive. Thoughts?

File: source/BankPanel.cpp [hunk 26]
```diff
@@ -24,6 +24,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "GameData.h"
 #include "Information.h"
 #include "Interface.h"
+#include "Mortgage.h"
```
Comment (Koranir) [inline source/BankPanel.cpp:27]:
```suggestion
```

File: source/Account.h [hunk 27]
```diff
@@ -21,75 +21,183 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
+	// A list containing mortgages taken out and fines assigned to the player.
+	std::vector<Mortgage> mortgages;
+
+	// If back salaries cannot be paid, it piles up rather than being ignored.
+	int64_t overdueCrewSalaries = 0;
+
+	// If back maintenance cannot be paid, it piles up rather than being ignored.
+	int64_t overdueMaintenance = 0;
+
+	// Regular income from salaries paid to the player.
```
Comment (Koranir) [inline source/Account.h:201]: Might as well explain what we're mapping to what, while we're at it here.

File: source/Account.h [hunk 28]
```diff
@@ -21,75 +21,183 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include <map>
 #include <optional>
 #include <string>
+#include <tuple>
```
Comment (Koranir) [inline source/Account.h:24]:
```suggestion
```

File: source/Account.h [hunk 29]
```diff
@@ -21,75 +21,183 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
+	// Access the map of player salaries.
+	const std::map<std::string, int64_t> &SalariesIncome() const;
+
+	// Return the sum of all player salaries.
+	int64_t SalariesIncomeTotal() const;
+
+	// Set the number of credits in a specified salary. This will add the
+	// salaries if it does not already exist, or modify and existing salary.
+	// If the amount passed in is 0, then the salary will be removed.
+	void SetSalariesIncome(const std::string &name, int64_t amount);
```
Comment (Koranir) [inline source/Account.h:121]:
```suggestion
	void SetSalaryIncome(const std::string &name, int64_t amount);
```

File: source/Account.cpp [hunk 30]
```diff
@@ -150,322 +219,448 @@ void Account::PayExtra(int mortgage, int64_t amount)
...
+		[](int64_t value, const std::map<string, int64_t>::value_type &salary)
+		{
+			return value + salary.second;
+		}
+	);
+}
+
+
+
+void Account::SetSalariesIncome(const string &name, int64_t amount)
```
Comment (Koranir) [inline source/Account.cpp:327]:
```suggestion
void Account::SetSalaryIncome(const string &name, int64_t amount)
```

**Review by Koranir — CHANGES_REQUESTED** [review]
You still have a few other reviews that haven't been addressed yet that should also be.

**Comment by petervdmeer** [comment]
Closing this PR; The large review effort on this PR is far out of balance for the minor readability benefits that a refactor of this class would bring, and there are still open review comments after 5 reminders to handle those open review comments.

**Comment by Anarchist2** [comment]
Seeing that the primary reason for closing this PR is a lack of reviewers on our end, I have elected to reopen this PR. However, do not take this as a guarantee that this PR won't be closed for other reasons, nor an indication that we will be able to review this any time soon.

I will also reiterate the review comments that have still not been addressed from before Koranir's review on the 24th:
https://github.com/endless-sky/endless-sky/pull/9484/files#r1438351175
https://github.com/endless-sky/endless-sky/pull/9484/files#r1438353995
https://github.com/endless-sky/endless-sky/pull/9484/files#r1672834201

**Comment by RisingLeaf** [comment]
Added three labels, it is waiting on OP in 7 comments, peter needs to respond in about 4 iirc. And based on comments by the author there is no intention of picking back up

File: source/Account.cpp (hunk 24 above)
Comment (petervdmeer) [reply source/Account.cpp:545]: No.

If we leave the utility function without using it, then it would just sit there like dead code. I also don't expect it to be used later down the line because of the "mental switching" as mentioned before, and if we ever do want to use it, then we could easily implment it using 1 line of code (using something like [erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2).

File: source/Account.cpp (hunk 25 above)
Comment (petervdmeer) [reply source/Account.cpp:528]: Both options 1 and 2 don't sound great, and I don't know if my suggestion from July 10th to add a string to the Receipt struct itself (so that the Receipt would internally have the information on what it is for) would work.

Magic numbers are still not okay here, because they seem to make the code quite a bit harder to read.

Please try to find a solution that is easier to read (and thus more maintainable).

**Comment by petervdmeer** [comment]
> Seeing that the primary justification for closing this PR was a lack of reviewing capability, as opposed to any inherent fault in the PR, I have reopened it.

My previous comment on this PR sounded a bit rude. Sorry for that.

This PR introduces some new code which seems less readable and less maintainable than the original code. That is just not okay for a refactoring PR, because refactoring by definition should make the code better than the original code.

Reviews were not lacking though; I did review the code already 8 times (most reviews at direct request of the PR author), the first review feedback on the reduced readability was already posted on Nov 23, 2023 and this PR still has open review-comments from Dec 29, 2023. 

It's nothing personal, and I really want people to contribute to the game, but I don't see how this PR is going to a mergeable state if more than 8 reviews and more than 10 months are not sufficient to address the readability issues and open review comments.
(And that is not to blame the PR author; things like complexity of the codebase, undocumented expectations, complexity of the GitHub-UI, possible unclear communication from me or other reviewers, or other reasons could easily play a role in this PR not getting to a mergeable state.)

> Added three labels, it is waiting on OP in 7 comments, peter needs to respond in about 4 iirc. And based on comments by the author there is no intention of picking back up

I replied to the latest open comments.

**Comment by shitwolfymakes** [comment]
To start, I want to give props for recognizing your overreaction and remediating it. Great job. End of praise.

As to the repeated characterization of my proposed code as "less readable and less maintainable", these are objectively incorrect by every metric and I have stated with numbers to back it up. At a certain point, thinking that following the logic of a function that *gasp* calls other functions is difficult even to an inexperienced developer, is a problem resting with the person who thinks that. So lets leave the Step function at 170 lines. Hell, lets put the whole game in a single fucking function, because you're right. Following the control flow into another function takes an amount of effort.

This is only the symptom of a larger problem. You are correct in that this PR has been open for 10 months, now a year. After the initial round of corrections, that were necessary because I hadn't accounted for the style guide, the feedback changed. It turned into constantly moving the goalposts and deafening silence for months at a time. I am not the only contributor who has noticed this, but I am the one most willing to call bullshit as I see it. Consider the following a trip to the woodshed, because I fucking love this game and despite the new content, it's in its sorriest state since MZ went incommunicado and we could push steam updates.

I see no evidence that the dev team has any interest in making noticeable change to the engine, or at least in making changes they personally disagree with. The boarding rework attempts (#6986) is the clearest and oldest example I've seen, followed by this PR. It has been kicked around for years because the devs either don't want it or don't care. 

As a player of the game since *2016*, whose favorite part is building absurd doomstacks that crater the framerate and push the engine to it's absolute limit, this is an infuriating and familiar pattern. Rather than accept a change that some players have been begging for, *for years*, and would leave other players who don't engage with that mechanic completely unaffected (*"but the missions that require boarding!"* is not a fucking excuse, it is trivial to generate custom ships/fleets), the devs hem and haw and equivocate and generally do anything other than accept community input on anything except the parts they care about.

At this point, I will bet money that if a PR was opened to migrate the existing style guide to an auto-linter CI or pre-commit hook, the existing dev team would immediately dismiss it out of hand because "that's harder to understand" than double-checking every fucking change you make against a webpage by hand, and then descending into a nitpicking spiral in the PR comments. Earlier this year I cancelled my own tool for the game, [Endless Sky Mission Builder](https://github.com/shitwolfymakes/Endless-Sky-Mission-Builder). I was the second person to take a major shot at building one, and I got much further than the original. It was made clear to me that there was no interest in changing from the outdated, error-prone, custom data format to something serializable that could be validated programmatically, even if I did it myself. 


### TL;DR
If you're reading this and considering contributing -- don't. Not right now. The devs don't care about community feedback and they don't care about making positive changes to the game engine. All they care about is making the changes they want, and damn the rest of us. And this is a pattern. There was a large group of devs and content creators who *forked the fucking game* a few years back because they were sick of this shit.

When the existing dev team is either gone or pull their heads out of their asses, I will consider *ever* contributing to this game again in a way that isn't hours played in my steam library. I've muted the ESC Discord, so DM me if you read this closing comment and want to talk.
